### PR TITLE
ExPlat: Self-Contained Client Internals (except for request)

### PR DESCRIPTION
**Pull request by jessie-ross**
**This PR adds internal functions for the Explat Self-contained client.**
- Just missing `request.ts`

Part of a series rebuilding the reference WIP PR #48475

#### Testing instructions

* See added Automated Tests 
* `yarn test-packages explat-client`

### CODE REVIEW

File: packages/explat-client/src/internal/experiment-assignments.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/**
+ * Internal dependencies
+ */
+import { ExperimentAssignment } from '../types';
```
Comment (sarayourfriend) [inline packages/explat-client/src/internal/experiment-assignments.ts]: If you're importing just a type you can use `import type` to help speed up type erasure.

File: packages/explat-client/src/internal/experiment-assignments.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/**
+ * Internal dependencies
```
Comment (yanirs) [inline packages/explat-client/src/internal/experiment-assignments.ts:2]: It took me a while to understand that `Internal dependencies` references to the imports, not the entire file. In general, the first block comment usually describes the file, so this is pretty confusing. How about we change these to be more descriptive and make the internal/external dependencies one-line comments?

Taking this file as an example, it'd be:

```javascript
/**
 * Functions to manipulate and check ExperimentAssignment objects.
 */

// Internal dependencies
import type { ExperimentAssignment } from '../types';
...
```

File: packages/explat-client/src/internal/state.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/**
+ * Internal dependencies
+ */
+import type { ExperimentAssignment } from '../types';
+import * as Validations from './validations';
+import * as ExperimentAssignments from './experiment-assignments';
+
+const experimentNameToExperimentAssignment: Record< string, ExperimentAssignment | undefined > = {};
+
+export function storeExperimentAssignment( experimentAssignment: ExperimentAssignment ): void {
```
Comment (yanirs) [inline packages/explat-client/src/internal/state.ts:15]: It'd be good to document these functions too (generally: anything that's exported).

File: packages/explat-client/src/internal/test-common.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export const validExperimentAssignmentA = {
```
Comment (yanirs) [inline packages/explat-client/src/internal/test-common.ts]: Should this file also be under `internal/test`?

File: packages/explat-client/src/internal/test/experiment-assignments.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
+/**
+ * Internal dependencies
+ */
```
Comment (yanirs) [inline packages/explat-client/src/internal/test/experiment-assignments.ts:3]: I think we can drop these headers from test files.

File: packages/explat-client/src/internal/test-common.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+	experimentName: 'experiment_name_a',
+	variationName: 'treatment',
+	retrievedTimestamp: Date.now(),
+	ttl: 60,
+};
+
+export const validExperimentAssignmentB = {
```
Comment (yanirs) [inline packages/explat-client/src/internal/test-common.ts]: Maybe call this `validFallbackExperimentAssignment`?

File: packages/explat-client/src/internal/test/state.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
+/**
+ * Internal dependencies
+ */
+import * as State from '../state';
+import { validExperimentAssignmentA, validExperimentAssignmentB } from '../test-common';
+
+describe( 'state', () => {
+	it( 'should save and retrieve valid EAs', () => {
```
Comment (yanirs) [inline packages/explat-client/src/internal/test/state.ts]: Let's spell out `EA` or simply use the word "assignment" where it's clear what's meant from the context.

File: packages/explat-client/package.json [hunk 8]
```diff
@@ -21,5 +21,8 @@
 		"watch": "tsc --build ./tsconfig.json --watch",
 		"prepack": "yarn run clean && yarn run build",
 		"test": "yarn jest"
+	},
+	"dependencies": {
+		"@automattic/calypso-polyfills": "^1.0.0"
```
Comment (yanirs) [inline packages/explat-client/package.json:26]: Should this be a dev dependency since it's only used by tests? Or will it be used by non-test files?

**Review by yanirs — APPROVED** [review]
LGTM. Reviewing the reference PR definitely helped. :slightly_smiling_face: 

See inline comments for mostly cosmetic issues.

File: packages/explat-client/src/internal/experiment-assignments.ts (hunk 2 above)
Comment (sarayourfriend) [reply packages/explat-client/src/internal/experiment-assignments.ts:2]: For what it's worth, this is a convention enforced by our ESLint rules and is followed by both `wp-calypso` and WordPress/Gutenberg.

File: packages/explat-client/src/internal/test-common.ts (hunk 4 above)
Comment (jessie-ross) [reply packages/explat-client/src/internal/test-common.ts]: I tried to get it to work like that but the conventions don't allow it - they don't allow exports.

File: packages/explat-client/src/internal/experiment-assignments.ts (hunk 2 above)
Comment (jessie-ross) [reply packages/explat-client/src/internal/experiment-assignments.ts:2]: Yep, they are compulsory if we are keeping with the conventions. I do think the files are self-explanatory enough and are pretty simple, plus we will have a little file guide in the readme.

File: packages/explat-client/package.json (hunk 8 above)
Comment (jessie-ross) [reply packages/explat-client/package.json:26]: Ah yep it should be devDep, I thought I checked though 😮

File: packages/explat-client/src/internal/experiment-assignments.ts (hunk 2 above)
Comment (yanirs) [reply packages/explat-client/src/internal/experiment-assignments.ts:2]: Makes sense.

**Pull request merged**